perses/rjmc/atom_mapping.py

```diff
@@ -0,0 +1,1475 @@
+
+        Parameters
+        ----------
+        old_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The initial molecule for the transformation.
+        new_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The final molecule for the transformation.
+
+        Returns
+        -------
+        atom_mappings : list of AtomMapping
+            All valid atom mappings
+
+        Examples:
+        ---------
+        Specify two molecules without positions
+
+        >>> from openff.toolkit.topology import Molecule
+        >>> ethane = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])([H:7])')
+        >>> ethanol = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])[O:7][H:8]')
+
+        Retrieve all mappings between two molecules
+
+        >>> atom_mappings = atom_mapper.get_all_mappings(ethane, ethanol)
+
+        You can also specify OEMols for one or more of the molecules:
+
+        >>> atom_mappings = atom_mapper.get_all_mappings(ethane.to_openeye(), ethanol.to_openeye())
+        >>> atom_mappings = atom_mapper.get_all_mappings(ethane.to_openeye(), ethanol)
+
+        """
+        atom_mappings = set() # all unique atom mappings found
+
+        # Create OpenFF and OEMol copies of input molecules, retaining properties if provided
+        # TODO: Do we really need to retain OEMol atom/bond inttypes?
+        def copy_molecule(mol):
+            """Generate OpenFF and OEMol copies of a given molecule,
+            retaining OEMol properties (such as atom/bond inttypes) if present.
+
+            Parameters
+            ----------
+            mol : openeye.oechem.OEMol or openeye.oechem.OEGraphMol or openff.toolkit.topology.Molecule
+                Molecule to be copied
+
+            Returns
+            -------
+            offmol : openff.toolkit.topology.Molecule
+                The OpenFF Molecule
+            oemol : openeye.oechem.OEMol
+                A copy of the OEMol, preserving data if provided
+            """
+            from openff.toolkit.topology import Molecule
+            from openeye import oechem
+            if isinstance(mol, Molecule):
+                # OpenFF Molecule
+                offmol = Molecule(mol)
+                oemol = offmol.to_openeye()
+                # Fix atom name retrieval from OEMol objects
+                # TODO: When https://github.com/openforcefield/openff-toolkit/issues/1026 is fixed, remove this
+                for atom, oeatom in zip(offmol.atoms, oemol.GetAtoms()):
+                    atom.name = oeatom.GetName()
+            elif hasattr(mol, 'GetAtoms'):
+                # OpenEye OEMol
+                offmol = Molecule(mol)
+                oemol = oechem.OEMol(mol)
+            else:
+                raise ValueError(f'Not sure how to copy {mol}; should be openff.toolkit.topology.Molecule or openeye.oechem.OEMol')
+
+            return offmol, oemol
```

Comment:
Not sure how bad we are in memory usage or if we would ever need a better memory footprint. But this might be a candidate for improvement, maybe returning just copies of the same type as the input. That is, if input is an `OEMol` output is just `OEMol` as well (similar with `offmol`).

Reply:
Great point! The hope is to eliminate `OEMol` objects entirely when possible, and just use `Molecule` in the future. This is just a temporary hack.